File: Data Structures & Algorithms/longest-increasing-subsequence/submission-6.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        LIS = [1] * n

        for i in range(n - 1, -1, -1):
            for j in range(i + 1, n):
                if nums[i] < nums[j]:
                    LIS[i] = max(LIS[i], 1 + LIS[j])
        
        return max(LIS)

        # memo = {}

        # def dfs(last, i):
        #     if i >= len(nums):
        #         return 0
        #     if (last, i) not in memo:
        #         if last == -1 or nums[last] < nums[i]:
        #             memo[(last, i)] = max(1 + dfs(i, i + 1), dfs(last, i + 1))
        #         else:
        #             memo[(last, i)] = dfs(last, i + 1)

        #     return memo[(last, i)]
        
        # return dfs(-1, 0)

File: Data Structures & Algorithms/house-robber-ii/submission-6.cpp
class Solution {
public:
    unordered_map<string, int> dp;

    int rob(vector<int>& nums) {
        if (nums.size() == 1) {
            return nums[0];
        }
        return max(dfs(nums, 0, nums.size() - 1), dfs(nums, 1, nums.size()));
    }
private:
    int dfs(vector<int>& nums, int i, int end) {
        if (i >= end) {
            return 0;
        }
        if (!dp[to_string(i) + "," + to_string(end)]) {
            dp[to_string(i) + "," + to_string(end)] = max(nums[i] + dfs(nums, i + 2, end), dfs(nums, i + 1, end));
        }
        return dp[to_string(i) + "," + to_string(end)];
        
    }

};
